crates/walrus-sdk/src/client/quilt_client.rs

```diff
@@ -0,0 +1,583 @@
+        } else {
+            return Err(ClientError::from(ClientErrorKind::Other(
+                format!("Path is neither a file nor a directory: {:?}.", path).into(),
+            )));
+        }
+    }
+
+    let mut collected_files_with_content: Vec<(PathBuf, Vec<u8>)> = Vec::new();
+    for file_path in collected_files {
+        let content = fs::read(&file_path).map_err(ClientError::other)?;
+        collected_files_with_content.push((file_path, content));
+    }
+
+    Ok(collected_files_with_content)
+}
+
+/// Get all file paths from a directory recursively.
+fn get_all_files_from_dir<P: AsRef<Path>>(path: P) -> ClientResult<Vec<PathBuf>> {
+    let path = path.as_ref();
+    let mut collected_files: HashSet<PathBuf> = HashSet::new();
+    let dir_entries = fs::read_dir(path).map_err(ClientError::other)?;
+    for entry in dir_entries {
+        let current_entry_path = entry.map_err(ClientError::other)?.path();
+        if current_entry_path.is_file() {
+            collected_files.insert(current_entry_path);
+        } else if current_entry_path.is_dir() {
+            collected_files.extend(get_all_files_from_dir(&current_entry_path)?);
+        }
+    }
+
+    Ok(collected_files.into_iter().collect())
+}
+
+/// A facade for interacting with Walrus quilt.
+#[derive(Debug, Clone)]
+pub struct QuiltClient<'a, T> {
+    client: &'a Client<T>,
+}
+
+impl<'a, T> QuiltClient<'a, T> {
+    const MAX_RETRIEVE_SLIVERS_ATTEMPTS: usize = 2;
+    const TIMEOUT_DURATION: Duration = Duration::from_secs(10);
+
+    /// Creates a new QuiltClient.
+    pub fn new(client: &'a Client<T>) -> Self {
+        Self { client }
+    }
+}
+
+impl<T: ReadClient> QuiltClient<'_, T> {
+    /// Retrieves the [`QuiltMetadata`].
+    ///
+    /// If not enough slivers can be retrieved for the quilt index, the entire blob will be read.
+    pub async fn get_quilt_metadata(&self, quilt_id: &BlobId) -> ClientResult<QuiltMetadata> {
+        self.client.check_blob_id(quilt_id)?;
+        let (certified_epoch, _) = self
+            .client
+            .get_blob_status_and_certified_epoch(quilt_id, None)
+            .await?;
+        let metadata = self
+            .client
+            .retrieve_metadata(certified_epoch, quilt_id)
+            .await?;
+
+        // Try to retrieve the quilt index from the slivers.
+        let quilt_index =
+            if let Ok(quilt_index) = self.retrieve_quilt_index(&metadata, certified_epoch).await {
+                quilt_index
+            } else {
+                // If the quilt index cannot be retrieved from the slivers, try to retrieve the
+                // quilt.
+                tracing::debug!(
+                    "failed to retrieve index slivers, trying to get quilt instead {}",
+                    quilt_id
+                );
+                self.get_quilt_enum(&metadata, certified_epoch)
+                    .await?
+                    .get_quilt_index()?
```

Comment:
Important: If we download the full quilt here, we should cache it and use it later to extract the patches.